### ISSUE TITLE
Race on gateway with etcd: STS token can arrive before etcd watch event makes it valid

### ISSUE THREAD

**Issue by adamhooper** [minio/minio#11149]
Immediately after a minio gateway stores an STS token in etcd, that token isn't usable from other minio gateways on the same etcd cluster.

## Expected Behavior

I'd expect a returned STS token to be readable by each minio gateway that connects to the etcd cluster.

## Current Behavior

Currently, each minio server "watches" for new etcd keys. minio returns an STS token once the write succeeds -- _not_ once the STS token is usable on all minio clients using the etcd cluster.

When a client requests and uses an STS token faster than the latency minio+etcd clusters, the gateway returns `InvalidAccessKeyId` even though the STS token is valid.

In our cluster, where we CPU-throttle minio and etcd, this occurs commonly. Judging by user reports and our own tests, maybe 5% of STS-token usages.

## Possible Solution

Three ways to change STS in gateway mode:

* Nix the local IAM cache. To validate an STS token, query etcd for the token's account ID, with `?quorum=true` to avoid etcd races
* Nix the local IAM cache. When creating an STS token, encode etcd's response's `modifiedIndex` into the token itself (hand-wavy here, there might be a chicken-and-egg problem). To validate an STS token, query etcd (without necessarily adding `?quorum=true`), and if the response's `X-Etcd-Index` is too low, set up a brief `?wait=[X-Etcd-Index]` request to wait for that etcd node to catch up.
* Implement `globalNotificationSys`, so the solution from server (that worked to solve #7991) works here. This means all minio nodes must be able to talk to each other.

## Steps to Reproduce (for bugs)

```bash

# 1. Set up our "Fake S3"
docker network create minio-test
docker run -d --network minio-test --name minio-test-s3 --cpus=0.5 \
    --env MINIO_ACCESS_KEY=root \
    --env MINIO_SECRET_KEY=rootroot \
    minio/minio:RELEASE.2020-12-18T03-27-42Z \
    server /data

# 2. Set up an etcd cluster: 3 nodes. Give them high latency.
for i in `seq 1 3`; do
    docker run -d --rm --network minio-test --name minio-test-etcd-$i \
        --cap-add=NET_ADMIN \
        gcr.io/etcd-development/etcd:v3.4.14 \
        /usr/local/bin/etcd \
        --name minio-test-etcd-$i \
        --initial-advertise-peer-urls http://minio-test-etcd-$i:2380 \
        --listen-peer-urls http://0.0.0.0:2380 \
        --advertise-client-urls http://minio-test-etcd-$i:2379 \
        --listen-client-urls http://0.0.0.0:2379 \
        --initial-cluster minio-test-etcd-1=http://minio-test-etcd-1:2380,minio-test-etcd-2=http://minio-test-etcd-2:2380,minio-test-etcd-3=http://minio-test-etcd-3:2380 \
        --initial-cluster-state new \
        --initial-cluster-token minio-test

  # add 50ms latency to etcd's egress
  docker exec minio-test-etcd-$i tc qdisc add dev eth0 root netem delay 50ms
done

# 3. Set up a minio gateway cluster: 3 nodes, each connected to its own etcd node
# Make them visible at localhost:9001, localhost:9002, localhost:9003
for i in `seq 1 3`; do
    docker run -d --network minio-test --name minio-test-minio-$i --cpus=0.5 \
        --env MINIO_ETCD_ENDPOINTS=http://minio-test-etcd-$i:2379 \
        --env MINIO_ACCESS_KEY=root \
        --env MINIO_SECRET_KEY=rootroot \
        --env AWS_ACCESS_KEY_ID=root \
        --env AWS_SECRET_ACCESS_KEY=rootroot \
        -p 900$i:9000 \
        minio/minio:RELEASE.2020-12-18T03-27-42Z \
        gateway s3 http://minio-test-s3:9000
done

# 4. Create user, "guest:guestguest", on etcd, that we'll use for STS AssumeRole
docker run -it --rm --network minio-test --entrypoint sh \
    minio/mc:RELEASE.2020-12-18T10-53-53Z \
    -c 'mc config host add main http://minio-test-minio-1:9000 root rootroot && mc admin user add main guest guestguest && mc admin policy set main readwrite user=guest'

# 5. Start Python -- we'll use it as our client
python
```

And in Python:

```python
import boto3
import time

def attempt(sts_endpoint: str, s3_endpoint: str):
    sts_client = boto3.session.Session(
        aws_access_key_id='guest',
        aws_secret_access_key='guestguest',
        region_name='us-east-1',
    ).client('sts', endpoint_url=sts_endpoint)
    credentials = sts_client.assume_role(
        RoleArn='notusednotusednotused',
        RoleSessionName='notused',
    )['Credentials']
    print("minio-sts %s => AccessKey %s => minio-gateway %s..." % (sts_endpoint, credentials['AccessKeyId'], s3_endpoint))
    # time.sleep(2)
    # S3 request goes to a different minio server
    s3_client = boto3.session.Session(
        aws_access_key_id=credentials['AccessKeyId'],
        aws_secret_access_key=credentials['SecretAccessKey'],
        aws_session_token=credentials['SessionToken'],
        region_name='us-east-1',
    ).client('s3', endpoint_url=s3_endpoint)
    s3_client.list_buckets()  # Should not error with InvalidAccessKeyId

for i in range(100):
    attempt(f"http://localhost:900{i % 3 + 1}", f"http://localhost:900{(i + 1) % 3 + 1}")
```

Results:

```
minio-sts http://localhost:9001 => AccessKey SL1FC9MN89UTL6RUOBKK => minio-gateway http://localhost:9002...
minio-sts http://localhost:9002 => AccessKey VWPFNGCC47NSDKGWNMJM => minio-gateway http://localhost:9003...
minio-sts http://localhost:9003 => AccessKey 4AY68TJ8OS6DJSHMYEOZ => minio-gateway http://localhost:9001...
minio-sts http://localhost:9001 => AccessKey F1FYG44C48CL8D0LRWJ2 => minio-gateway http://localhost:9002...
minio-sts http://localhost:9002 => AccessKey R9KW8TC39DQQXIZGDA16 => minio-gateway http://localhost:9003...
minio-sts http://localhost:9003 => AccessKey L8W0QROH7CJV26WJYMZY => minio-gateway http://localhost:9001...
minio-sts http://localhost:9001 => AccessKey F91N3EUT5FOVIKO95CJB => minio-gateway http://localhost:9002...
minio-sts http://localhost:9002 => AccessKey 1UCIZWF7VH8CPSETXKWT => minio-gateway http://localhost:9003...
minio-sts http://localhost:9003 => AccessKey WI899F3TH7F2AFUK5AJV => minio-gateway http://localhost:9001...
minio-sts http://localhost:9001 => AccessKey 168Y8I2J8VN7W5LACYYY => minio-gateway http://localhost:9002...
Traceback (most recent call last):
  File "<stdin>", line 2, in <module>
  File "<stdin>", line 20, in attempt
  File "/home/adam/.local/lib/python3.9/site-packages/botocore/client.py", line 357, in _api_call
    return self._make_api_call(operation_name, kwargs)
  File "/home/adam/.local/lib/python3.9/site-packages/botocore/client.py", line 676, in _make_api_call
    raise error_class(parsed_response, operation_name)
botocore.exceptions.ClientError: An error occurred (InvalidAccessKeyId) when calling the ListBuckets operation: The Access Key Id you provided does not exist in our records.
```

The latency has a big effect. I've replicated with etcd egress latency=30ms, not with 0ms.

## Context

We use STS to manage file upload. We have a 3-node etcd cluster and 2 minio gateways, all in Kubernetes.

In our cluster, where we CPU-throttle minio and etcd, this occurs commonly. Judging by user reports, maybe 5% of STS-token usages.

We throttle minio and etcd so we can use CPU for Other Things. Maybe that exacerbates the problem?

## Regression

This is not a regression.

I previously filed #7991, but my steps to reproduce actually missed _this_ bug and triggered a _different_ bug. And that's the one that was fixed. Oops.

## Your Environment

* `gcr.io/etcd-development/etcd:v3.4.14`
* `minio/minio:RELEASE.2020-12-18T03-27-42Z`
* Docker CE 19.03.13; Kubernetes v1.16.15-gke.4300

**Comment by harshavardhana** [comment]
> * Implement `globalNotificationSys`, so the solution from server (that worked to solve #7991) works here. This means all minio nodes must be able to talk to each other.

Gateway's are share nothing they cannot implement a common notification system, this is why we use etcd in the first place for common shared location. 

So it is not possible to practically extend gateway in this manner. 

> In our cluster, where we CPU-throttle minio and etcd, this occurs commonly. Judging by user reports, maybe 5% of STS-token usages.

Yes, this is perhaps why it would slow down etcd response time, if there is a timing difference we can of course try to read from etcd directly without waiting for the watch to be invoked asynchronously. 

Let me see if we add some ways to avoid this problem add a fallback.

**Comment by harshavardhana** [comment]
Please test https://github.com/minio/minio/pull/11151 @adamhooper - can't reproduce anymore. 

```
minio-sts http://localhost:9001 => AccessKey 2FGZ74O6FA03H59DKAP4 => minio-gateway http://localhost:9002...
minio-sts http://localhost:9002 => AccessKey TUGZ333S17EALSFP665I => minio-gateway http://localhost:9003...
minio-sts http://localhost:9003 => AccessKey XG14OOIE26VUN67FBYEV => minio-gateway http://localhost:9001...
minio-sts http://localhost:9001 => AccessKey HUQNBZ417V2QO7DK9MKC => minio-gateway http://localhost:9002...
minio-sts http://localhost:9002 => AccessKey SQVI068W7Y0C6ALAPM4X => minio-gateway http://localhost:9003...
minio-sts http://localhost:9003 => AccessKey PKQ3LO7TMSA11OLNVX3P => minio-gateway http://localhost:9001...
minio-sts http://localhost:9001 => AccessKey RL8RCYETWYYP7EFT9W04 => minio-gateway http://localhost:9002...
minio-sts http://localhost:9002 => AccessKey NKV11TKPTFIVQKXI34C7 => minio-gateway http://localhost:9003...
minio-sts http://localhost:9003 => AccessKey 3ZIR4T3U6G073Q6R01HF => minio-gateway http://localhost:9001...
minio-sts http://localhost:9001 => AccessKey BDPLKQU69B0C62J6XGLU => minio-gateway http://localhost:9002...
..
```

**Comment by adamhooper** [comment]
@harshavardhana I can still reproduce on the `fix-iam` branch.

Here are my tweaks to the script that make it happen:

* Added `--cpus=0.25` to etcd (I suspect this makes the bug easier to reproduce)
* Ran `minio:fix-iam`, which I built using `docker build -t minio:fix-iam` from a checked-out `harshavardhana/fix-iam` branch (commit `95289934b221390a2cab7afa449201158b913ea0`)

```
# 1. Set up our "Fake S3"
docker network create minio-test
docker run -d --network minio-test --name minio-test-s3 --cpus=0.5 \
    --env MINIO_ACCESS_KEY=root \
    --env MINIO_SECRET_KEY=rootroot \
    minio/minio:RELEASE.2020-12-18T03-27-42Z \
    server /data

# 2. Set up an etcd cluster: 3 nodes. Give them high latency and slow CPUs
for i in `seq 1 3`; do
    docker run -d --rm --network minio-test --name minio-test-etcd-$i \
        --cap-add=NET_ADMIN \
        --cpus=0.25 \
        gcr.io/etcd-development/etcd:v3.4.14 \
        /usr/local/bin/etcd \
        --name minio-test-etcd-$i \
        --initial-advertise-peer-urls http://minio-test-etcd-$i:2380 \
        --listen-peer-urls http://0.0.0.0:2380 \
        --advertise-client-urls http://minio-test-etcd-$i:2379 \
        --listen-client-urls http://0.0.0.0:2379 \
        --initial-cluster minio-test-etcd-1=http://minio-test-etcd-1:2380,minio-test-etcd-2=http://minio-test-etcd-2:2380,minio-test-etcd-3=http://minio-test-etcd-3:2380 \
        --initial-cluster-state new \
        --initial-cluster-token minio-test

  # add 50ms latency to etcd's egress
  docker exec minio-test-etcd-$i tc qdisc add dev eth0 root netem delay 50ms
done

# 3. Set up a minio gateway cluster: 3 nodes, each connected to its own etcd node
# Make them visible at localhost:9001, localhost:9002, localhost:9003
for i in `seq 1 3`; do
    docker run -d --network minio-test --name minio-test-minio-$i \
        --env MINIO_ETCD_ENDPOINTS=http://minio-test-etcd-$i:2379 \
        --env MINIO_ACCESS_KEY=root \
        --env MINIO_SECRET_KEY=rootroot \
        --env AWS_ACCESS_KEY_ID=root \
        --env AWS_SECRET_ACCESS_KEY=rootroot \
        -p 900$i:9000 \
        minio:fix-iam \
        gateway s3 http://minio-test-s3:9000
done

# 4. Create user, "guest:guestguest", on etcd, that we'll use for STS AssumeRole
docker run -it --rm --network minio-test --entrypoint sh \
    minio/mc:RELEASE.2020-12-18T10-53-53Z \
    -c 'mc config host add main http://minio-test-minio-1:9000 root rootroot && mc admin user add main guest guestguest && mc admin policy set main readwrite user=guest'

# 5. Start Python -- we'll use it as our client
python
```

**Comment by adamhooper** [comment]
etcd seems like an icky fit here.

Brainstorming another way to distribute STS tokens: The policy and expiry time are compressed, encrypted with a private key (a secret environment variable shared by all a minio cluster's gateways), and _embedded in the session token_ during `AssumeRole`. During request, the minio gateway decrypts, decompresses and validates the session token. No need for etcd; no need for minio gateways to communicate with one another.

I'm out of my depth; but there are hints in the AWS API that suggest this is how AWS does it:

* AWS IAM imposes a `PackedPolicySize` cap on compressed JSON policies. It's very strict. Ref: https://docs.aws.amazon.com/STS/latest/APIReference/API_AssumeRole.html#API_AssumeRole_RequestParameters
* AWS provides no way to revoke an STS token. (There's a way to revoke _all_ STS tokens for a given user: internally, AWS attaches a new poliicy to the _user_ -- not the tokens). Ref: https://docs.aws.amazon.com/IAM/latest/UserGuide/id_roles_use_revoke-sessions.html

**Comment by harshavardhana** [comment]
> etcd seems like an icky fit here.
> 
> Brainstorming another way to distribute STS tokens: The policy and expiry time are compressed, encrypted with a private key (a secret environment variable shared by all a minio cluster's gateways), and _embedded in the session token_ during `AssumeRole`. During request, the minio gateway decrypts, decompresses and validates the session token. No need for etcd; no need for minio gateways to communicate with one another.
> 
> I'm out of my depth; but there are hints in the AWS API that suggest this is how AWS does it:
> 
> * AWS IAM imposes a `PackedPolicySize` cap on compressed JSON policies. It's very strict. Ref: https://docs.aws.amazon.com/STS/latest/APIReference/API_AssumeRole.html#API_AssumeRole_RequestParameters
> * AWS provides no way to revoke an STS token. (There's a way to revoke _all_ STS tokens for a given user: internally, AWS attaches a new poliicy to the _user_ -- not the tokens). Ref: https://docs.aws.amazon.com/IAM/latest/UserGuide/id_roles_use_revoke-sessions.html

That is a different way to do it by using shared secrets but that is not something that can be done as a fix anymore. Gateway is frozen and not taking in anymore features..

AFAIK this is the best possible fix - even after a successful write to etcd if etcd says that it doesn't have the data even after re-reading it then there is nothing MinIO can do really.

**Comment by adamhooper** [comment]
> AFAIK this is the best possible fix - even after a successful write to etcd if etcd says that it doesn't have the data even after re-reading it then there is nothing MinIO can do really.

Would it be too hard to read with `quorum=true` when resolving a race? I think an etcd write only succeeds after the quorum writes the value. Ref: https://etcd.io/docs/v3.4.0/learning/api_guarantees/ So if the read has `quorum=true`, then the read value will be the write value or a later one (as it should be)

**Comment by harshavardhana** [comment]
> Would it be too hard to read with `quorum=true` when resolving a race? I think an etcd write only succeeds after the quorum writes the value. Ref: https://etcd.io/docs/v3.4.0/learning/api_guarantees/ So if the read has `quorum=true`, then the read value will be the write value or a later one (as it should be)

Let me check how to set this with etcd client @adamhooper

**Comment by harshavardhana** [comment]
Okay quorum=true is older v2 API requirement that seems to be not required with v3 @adamhooper - this is something fundamental here to etcd - there seems to be no way to provide - quorum=true anymore.

**Comment by harshavardhana** [comment]
@adamhooper is this still a problem at your end? - I would like to close this issue out. Because I don't think there is much we can do here - throttling services while putting the load on them is counter-intuitive what you want out of etcd in this situation.

The patchwork while it helps - further throttling seems to cause more problems.

**Comment by adamhooper** [comment]
@harshavardhana Thanks for the ping.

Our (temporary) workaround is to reduce the number of minio gateways to 1.

Medium-term, we'll migrate away from STS. (We only use it for client-side file uploads ... and luckily, we supply the client-side code ourselves.)

In sum: it seems to me like the best solution to this race is to stop using etcd and encrypt policies directly into shared tokens. I understand that's out of scope.

**Comment by harshavardhana** [comment]
Merged the PR #11151 FWIW the maximum MinIO can do in this case, but there are further throttling and delayed propagation in etcd there is no way to handle. Since etcd seems like is not a strictly consistent store.
